File: tools/capture_ybseqs_reference.m
function capture_ybseqs_reference(seqs_dir, out_json)
% capture_ybseqs_reference  Serialize real experiment sequences for the reader test.
%
% Walks matlab_new/YbSeqs, builds each sequence ENGINE-FREE, and writes its
% serialize() bytes (hex) to OUT_JSON so the pyctrl reader round-trip test
% (tests/test_ybseqs_roundtrip_live.py) can verify the byte-format reader against
% real production sequences. Records a per-sequence status (ok / skipped / error).
%
% SAFETY -- this only BUILDS + serialize()s the tree:
%   * serialize() never runs the deferred regBeforeStart/regAfterEnd callbacks
%     (camera/AWG/server setup), so building does not touch that hardware.
%   * SeqManager.override_tick_per_sec(1000) keeps the libnacs engine unloaded.
%   * A name DENYLIST skips sequences that drive AWG / SLM / AOD / camera /
%     picomotor hardware directly in their body, so we never even attempt them.
%   * Run it as a SEPARATE headless process (matlab -batch); it adds paths and
%     touches the SeqManager singleton only within its own process.
% Still: prefer a maintenance window. Build is engine-free and callback-free, but
% it is real sequence code -- run it when you would be comfortable building a shot.
%
%   capture_ybseqs_reference                 % default YbSeqs -> reference_ybseqs/
%   capture_ybseqs_reference(dir, out)       % explicit paths

    here = fileparts(mfilename('fullpath'));
    repo = fullfile(here, '..', '..');
    mn = fullfile(repo, 'matlab_new');
    if nargin < 1 || isempty(seqs_dir), seqs_dir = fullfile(mn, 'YbSeqs'); end
    if nargin < 2 || isempty(out_json)
        out_json = fullfile(here, '..', 'tests', 'reference_ybseqs', 'ybseqs_reference.json');
    end
    odir = fileparts(out_json);
    if ~exist(odir, 'dir'), mkdir(odir); end

    % --- real framework on the path (NOT +archived / .claude worktrees) ---
    for d = {mn, fullfile(mn,'lib'), fullfile(mn,'YbSeqs'), fullfile(mn,'YbSteps'), ...
             fullfile(mn,'YbScans'), fullfile(mn,'YbScans','scanConfig'), ...
             fullfile(mn,'YbRearrangement'), fullfile(mn,'YbFPGAWaveforms'), ...
             fullfile(mn,'YbExptCtrl')}
        if exist(d{1}, 'dir'), addpath(d{1}); end
    end

    % Sequences that drive hardware directly in-body -> never attempt to build.
    denylist = 'AWG|SLM|AOD|Orca|Pico|Rearrange|HandOver|Trigger|SingleMove|setAWG|[Mm]odulation';

    % Real production tick rate (config.yml: 1e12 == 1 ps). The synthetic
    % reference_list.m builders could use 1000 because their steps are ms-scale;
    % real sequences use us-scale steps that round to 0 ticks at 1000/s and are
    % rejected ("Time offset/length must be positive"). Engine stays unloaded.
    SeqManager.override_tick_per_sec(1e12);
    cleanup = onCleanup(@() SeqManager.override_tick_per_sec(0)); %#ok<NASGU>

    files = dir(fullfile(seqs_dir, '*.m'));
    parts = {};
    n_ok = 0;
    for i = 1:numel(files)
        name = files(i).name(1:end-2);   % strip .m
        if ~isempty(regexp(name, denylist, 'once'))
            parts{end+1} = entry_json(name, 'skip:hardware-driver', ''); %#ok<AGROW>
            continue;
        end
        [hexb, status] = try_build(name);
        if strcmp(status, 'ok'), n_ok = n_ok + 1; end
        parts{end+1} = entry_json(name, status, hexb); %#ok<AGROW>
    end
    out = ['[', strjoin(parts, ','), ']'];

    fid = fopen(out_json, 'w');
    fwrite(fid, out, 'char');
    fclose(fid);
    fprintf('wrote %d sequences (%d serialized ok) to %s\n', numel(files), n_ok, out_json);
end


function [hexb, status] = try_build(name)
    hexb = '';
    try
        fn = str2func(name);
        nin = nargin(fn);
        if nin == 0
            s = fn();
        elseif nin == 1
            s = fn(ExpSeq());           % takes a configured ExpSeq, returns it
        else
            status = sprintf('skip:nargin=%d', nin);
            return;
        end
        if ~isa(s, 'ExpSeq')
            status = 'skip:not-an-ExpSeq';
            return;
        end
        b = serialize(s);
        hexb = lower(sprintf('%02x', typecast(int8(b), 'uint8')));
        status = 'ok';
    catch e
        msg = regexprep(e.message, '[\"\\\r\n\t]', ' ');
        if numel(msg) > 160, msg = msg(1:160); end
        status = ['err:' msg];
    end
end


function s = entry_json(name, status, hexb)
    s = sprintf('{"name":"%s","status":"%s","bytes":"%s"}', name, status, hexb);
end

File: tools/capture_ybseqs_reference.py
"""Re-capture ``tests/reference_ybseqs/ybseqs_reference.json`` FROM pyctrl.

Pyctrl-side twin of the MATLAB ``capture_ybseqs_reference.m``. Since pyctrl is now the
authoritative runtime and the MATLAB byte oracle is only a reference for un-migrated scans,
the strict pyctrl<->MATLAB byte-equality (THE ONE RULE) is RELEASED for these sequences: this
tool captures pyctrl's OWN ``serialize()`` output as the regression reference (mirrors
``tools/capture_config_reference.py`` for the config oracle). ``test_ybseqs_build.py`` then
becomes a pyctrl self-consistency guard -- "did a build's bytes change unexpectedly?" -- rather
than a MATLAB-equivalence check. Engine-free (override_tick_per_sec), like the test.

Only the pyctrl-buildable ``status == "ok"`` sequences (the ``_SEQS`` list, mirroring
``tests/test_ybseqs_build.py``) are rebuilt; every other entry (skip/hardware-driver) is
preserved verbatim so the file shape and the other consumers (test_ybseqs_roundtrip) are
unaffected.

    python pyctrl/tools/capture_ybseqs_reference.py
"""

import json
import os
import sys

_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))   # .../pyctrl
for _p in (os.path.join(_ROOT, "lib"), os.path.join(_ROOT, "YbSteps"),
           os.path.join(_ROOT, "YbSeqs"), os.path.join(_ROOT, "YbExptCtrl"),
           os.path.join(_ROOT, "tools"), _ROOT):   # mirror pyproject.toml pythonpath
    if _p not in sys.path:
        sys.path.insert(0, _p)

import seq_manager                       # noqa: E402
from exp_seq import ExpSeq               # noqa: E402
from seq_config import SeqConfig         # noqa: E402

_REF = os.path.join(_ROOT, "tests", "reference_ybseqs", "ybseqs_reference.json")

# (name, nargin) -- MUST mirror tests/test_ybseqs_build.py::_SEQS (the file == module ==
# function name; nargin 0 -> the seq builds its own ExpSeq, 1 -> it takes one).
_SEQS = [
    ("CoreShellMOTSeq", 1),
    ("GreenMOTSeq", 1),
    ("DummySeq", 0),
    ("TweezerLoadingSeq", 1),
    ("BlueTweezerLoadingSeq", 1),
    ("TweezerEnhancedLoadingSeq", 1),
    ("CoolingOptimizationSeq", 1),
    ("ImagingSurvivalSeq", 1),
    ("ReleaseRecaptureSeq", 1),
    ("PushoutSurvivalSeq", 1),
    ("PushoutSurvival399Seq", 1),
    ("ImagingPushoutSurvivalSeq", 1),
    ("RearrangeCommSeq", 1),
    ("RearrangeCommSeq2", 1),
    ("get_my_seq", 1),
]


def _build(name, nargin):
    mod = __import__(name)                # file == module == function name (MATLAB name)
    fn = getattr(mod, name)
    return fn() if nargin == 0 else fn(ExpSeq())


def main():
    SeqConfig.reset()
    SeqConfig.load_real()
    seq_manager.override_tick_per_sec(1e12)   # us-scale steps need a fine tick (engine never loaded)
    try:
        with open(_REF) as f:
            entries = json.load(f)
        by_name = {e["name"]: e for e in entries}
        n_ok = 0
        for name, nargin in _SEQS:
            b = _build(name, nargin).serialize()
            e = by_name.get(name)
            if e is None:
                e = {"name": name}
                entries.append(e)
                by_name[name] = e
            e["status"] = "ok"
            e["bytes"] = b.hex()
            n_ok += 1
    finally:
        seq_manager.override_tick_per_sec(0)
        SeqConfig.reset()

    with open(_REF, "w") as f:
        json.dump(entries, f)
    print("re-captured %d 'ok' seqs from pyctrl -> %s" % (n_ok, _REF))


if __name__ == "__main__":
    main()
